Allow 'floyd data delete' to accept multiple ids

mckayward opened the pull request:
# Notes on these changes:

## No early exit for failed deletes
If one of the ids fails to delete, this approach still attempts to delete the other ids, but will have an exit status of 1. This is similar to docker's `rmi` approach to image deletion.

## Move logging and error handing for API calls into Client Classes
I didn't notice this when making the changes in #14, but issues #17 and #18 got me thinking about how to start handling failed API calls and their associated log messages more gracefully. Fundamentally, log messages and exception handling for requests to the API are the responsibility of the Client class (DataClient in this case), so I moved those tasks over there in these changes. I noticed that we are doing some logging in some Client classes already ([here](https://github.com/floydhub/floyd-cli/blob/master/floyd/client/data.py#L49) for example). The re-usability and consistency that gives is nice, and it cleans up the logic in the cli classes. Let me know what you think of this approach.

## ~~Don't GET a data ID before calling DELETE on it~~
~~I noticed that the only reason we get a data id before deleting it is so we can reference its name. We don't allow deletion by name anyways, so I thought it could be good to just remove reference to the name altogether. It's a bit confusing to tell the cli to delete something by id: `floyd data delete 12345`, and then have the cli tell you something like `Data username/data_name: Deleted`.~~

~~This somewhat awkward mix of name and id will be highlighted if we decide to continue to move toward handling API request logging and error handling in the Client classes, as some of the methods take only ids anyways (like `DataClient().delete(id)`, for example. Thoughts, @narenst? If we want to keep the old approach, it's an easy revert of my last commit.~~

narenst commented on floyd/client/data.py:
```diff
@@ -72,10 +74,19 @@ def create(self, data):
+        except ClickException as e:
+            floyd_logger.info("Data {}: {}".format(id, e.message))
+            return None
+        except HTTPError as e:
```
This class extends the `FloydHttpClient` which takes care of exception handling.

mckayward replied:
I'm not following...
`HTTPError` is what's raised by `raise_for_status()` ([see docs here](http://docs.python-requests.org/en/master/api/#requests.Response.raise_for_status)). The base client raises it [here](https://github.com/floydhub/floyd-cli/blob/master/floyd/client/base.py#L81). Am I missing something?

narenst commented on floyd/client/data.py:
```diff
@@ -72,10 +74,19 @@ def create(self, data):
+                                    "{}{}".format(self.url, id))
+            data_dict = response.json()
+            return Data.from_dict(data_dict)
+        except ClickException as e:
```
Curious, what is the flow which throws the `ClickException` here?

mckayward replied:
I used `ClickException` as a catch-all for the different Exceptions in [floyd.exceptions](https://github.com/floydhub/floyd-cli/blob/master/floyd/exceptions.py#L10-L31) that all subclass from it. These are the exceptions that `BaseClient` raises.

So, when an [exception is raised by `BaseClient`](https://github.com/floydhub/floyd-cli/blob/master/floyd/client/base.py#L57-L81), during a call to `request()`, it's either going to be a `ClickException` (from floyd.exceptions), or an [`HTTPError` raised by `request.raise_for_status()](https://github.com/floydhub/floyd-cli/blob/master/floyd/client/base.py#L81).

I think maybe the intention was to have the exceptions in [floyd.exceptions](https://github.com/floydhub/floyd-cli/blob/master/floyd/exceptions.py) sub-class from `FloydException`, but they actually sub from `ClickException`

narenst commented:
Agreed about the exception handling idea. There is already a [base client](https://github.com/floydhub/floyd-cli/blob/master/floyd/client/base.py) that has some exception handling. Maybe we can unify error handling (such as 500s) there?

narenst commented:
Re: using names vs ids. We want to eventually move away from ids and expose easy-to-use names to the end user. Eg; we want the user to be able to say `floyd delete johndoe/fast-style-transfer:2`. So I think we should continue to fetch the name for now.

mckayward commented:
> Re: using names vs ids. We want to eventually move away from ids and expose easy-to-use names to the end user. Eg; we want the user to be able to say floyd delete johndoe/fast-style-transfer:2. So I think we should continue to fetch the name for now.

Alright sounds good. I'll make the change back.

mckayward commented:
@narenst 
> Agreed about the exception handling idea. There is already a base client that has some exception handling. Maybe we can unify error handling (such as 500s) there?

Yeah I think we can flesh out the exception handling in the base client some more to better handle a wider range of exceptions. I'll think on this for a bit

mckayward commented:
> Agreed about the exception handling idea. There is already a base client that has some exception handling. Maybe we can unify error handling (such as 500s) there?

@narenst I think current design of the `FloydHttpClient` class is pretty good. Currently, the `FloydHttpClient().request()` method is flexible and allows the other clients to cleanly access the API.

If we just flesh out the exception list that `FloydHttpClient().request()` raises, have them all sub from `FloydException`, then each caller can choose to handle them in a way that's best in each situation. Usually this will just be passing along the default error message for that particular exception framed within the context of the call. However, sometimes the caller might want to do something more customized, and this approach allows the flexibility to do that.

As an example, I'm doing this with the [`DataClient().delete` method in this PR](https://github.com/mckayward/floyd-cli/blob/feature/multiple_data_delete/floyd/client/data.py#L98-L106). We ask the base client to make a delete request; if it succeeds, we print a `"Data 123: Deleted"` message; if it fails, we handle the exception and print the message of the exception that's raised, but in a format that is unique to the `DataClient`:  `"Data 123: ERROR! The resource you are looking for was not found. Check if the id is correct."` Other callers might want to handle certain exceptions differently, like by passing more customized error messages to the user, retrying a request, etc.

What do you think of this approach? I've implemented it in the most recent commit. (Probably with some tweaks needed to the additional Exceptions I added).

narenst commented on floyd/cli/data.py:
```diff
@@ -113,21 +114,34 @@ def output(id, url):
+                                   "Delete Data: {}?".format(data_source.name),
+                                   abort=False,
+                                   default=False
+                                 ):
```
For code style consistency, can you indent this as:
```
if not yes and not click.confirm("Delete Data: {}?".format(data_source.name),
                                 abort=False,
                                 default=False):
```

mckayward replied:
Done

narenst commented on floyd/client/base.py:
```diff
@@ -4,6 +4,13 @@
 from floyd.manager.auth_config import AuthConfigManager
 from floyd.exceptions import AuthenticationException, BadRequestException, NotFoundException, OverLimitException
+
+from floyd.exceptions import AuthenticationException, AuthorizationException, \
```
Seems like some of these are already imported above.

Also, can you use [parantheses](https://www.python.org/dev/peps/pep-0328/#rationale-for-parentheses) for grouping these imports? That is the pattern we use in the rest of the code.

mckayward replied:
Done

narenst commented on floyd/client/base.py:
```diff
@@ -69,13 +76,22 @@ def check_response_status(self, response):
+                raise BadGatewayException()
+            elif response.status_code == 504:
+                raise GatewayTimeoutException()
+            elif 500 <= response.status_code < 600:
```
Adding all the exception here in one place is great!

narenst commented on floyd/client/data.py:
```diff
@@ -72,19 +73,34 @@ def create(self, data):
+        except FloydException as e:
+            floyd_logger.info(
+               "Error while retrieving data: {}".format(e.message)
+            )
```
Minor: This line is not too long, so can we keep this in one line? (similar to the other functions in this class)

mckayward replied:
Done

narenst commented:
@mckayward The changes look good, except some minor comments.

It will be great to port these changes to the experiment client as well. Thanks!

mckayward commented:
@narenst Alright I made the style updates. Let me know if you spot anything else.

I'd be happy to get these changes ported to the experiment client. I'll send a PR for it.
